File: flaskr/static/amidaForm.js
const amidaForm          = document.querySelector("form");
const formElements       = document.querySelectorAll("input, select, textarea");

const errorDiv           = document.getElementById("errorDiv");

const titleInput         = document.getElementById("title");
const lineCountInput     = document.getElementById("line_count");
const amidaItemsTextarea = document.getElementById("items");
const optionAORadio      = document.getElementById("option_auto_open");
const optionHIRadio      = document.getElementById("option_hide_items");
const adminPwdInput      = document.getElementById("admin_password");
const adminPwdConfInput  = document.getElementById("admin_password_confirm");
const userPwdInput       = document.getElementById("user_password");
const userPwdConfInput   = document.getElementById("user_password_confirm");

const mode = amidaForm.dataset.mode;

const isASCII = str => /^[\x00-\x7F]*$/.test(str);

function clearError() {
    errorDiv.innerHTML = "";
}

function showError(msg) {
    errorDiv.textContent = msg;
}

function validateLineCount(lineCount) {
    lineCount = parseInt(lineCount, 10);
    return !Number.isNaN(lineCount) && lineCount >= 3 && lineCount <= 20

}

function validatePwd(pwd) {
    return isASCII(pwd) && pwd.length >= 3 && pwd.length <= 20
}

// フォームが送信（submit）される時の処理を登録
function validateAll() {
    clearError();

    // フォーム内容の取得
    const title        = titleInput.value.trim();
    const lineCount    = lineCountInput.value.trim();
    const amidaItems   = amidaItemsTextarea.value.trim();
    const adminPwd     = adminPwdInput.value.trim();
    const adminPwdConf = adminPwdConfInput.value.trim();
    const userPwd      = userPwdInput.value.trim();
    const userPwdConf  = userPwdConfInput.value.trim();

    // 検証：タイトル
    if (!title) {
        showError("タイトルは必ず入力して下さい");
        return false;
    }

    // 検証：本数（作成モードのみ）
    if (mode === "create") {
        if (!lineCount) {
            showError("本数を選択してください");
            return false;
        }
    }

    // 検証：アイテム
    if (!amidaItems) {
        showError("アイテムを入力してください");
        return false;
    }

    // 検証：管理者パスワード
    if (!adminPwd) {
        showError("管理パスワードを入力してください");
        return false;
    } else if (!validatePwd(adminPwd)) {
        showError("管理パスワードは 3 文字以上 20 文字以内で入力してください");
        return false;
    } else if (adminPwd !== adminPwdConf) {
        showError("管理パスワードは、上段と下段で同じものを入力して下さい");
        return false;
    }

    // 検証：利用パスワード（選択）
    if (userPwd) {
        if (!validatePwd(userPwd)) {
            showError("利用パスワードは 3 文字以上 20 文字以内で入力してください");
            return false;
        } else if (userPwd !== userPwdConf) {
            showError("利用パスワードは、上段と下段で同じものを入力して下さい");
            return false;
        }
    }

    return true;
}

formElements.forEach(element => {
    // 文字が入力されるたびにチェックを実行！
    element.addEventListener("input", () => {
        validateAll();
    });
});

amidaForm.addEventListener("submit", (event) => {
    // エラーがあったら送信止め
    if (!validateAll()) {
        event.preventDefault();
    }
});

function toggleFields() {
    // 本数正しい入力されなければ、下の枠を入力不可にする
    const lineCount = lineCountInput.value.trim();
    if (validateLineCount(lineCount)) {
        amidaItemsTextarea.disabled = false;
        optionAORadio.disabled = false;
        optionHIRadio.disabled = false;
        adminPwdInput.disabled = false;
        adminPwdConfInput.disabled = false;
        userPwdInput.disabled = false;
        userPwdConfInput.disabled = false;
    } else {
        amidaItemsTextarea.disabled = true;
        optionAORadio.disabled = true;
        optionHIRadio.disabled = true;
        adminPwdInput.disabled = true;
        adminPwdConfInput.disabled = true;
        userPwdInput.disabled = true;
        userPwdConfInput.disabled = true;
    }
}
toggleFields();
lineCountInput.addEventListener("input", toggleFields);

File: flaskr/amida/amida_form.py
"""あみだくじのフォーム制御"""

def validate_line_count(line_count):
    """線の本数の検証"""
    try:
        validated_count = int(line_count)
        if not 3 <= validated_count <= 20:
            raise ValueError("Invalid line_count")

    except (ValueError, TypeError):
        return False

    return True

def validate_password(password):
    """パスワードの検証"""
    return password.isascii() and 3 <= len(password) <= 20

def get_and_validate_form(form, mode, current_line_count=None):
    """ フォームの検証
    Return
        tuple: (errors, amida, amida_items)
    """

    # データの取得
    title = form.get("title", "").strip()
    title = title[:100]

    line_count = form.get("line_count")

    amida_items = form.get("items", "").splitlines()
    amida_items = [item.strip()[:20] for item in amida_items if item.strip()]

    option_auto_open = form.get("option_auto_open") == "on"
    option_hide_items = form.get("option_hide_items") == "on"

    admin_password = form.get("admin_password", "")
    admin_password_confirm = form.get("admin_password_confirm", "")

    user_password = form.get("user_password", "")
    user_password_confirm = form.get("user_password_confirm", "")

    # データの検証
    errors = {}
    if not title:
        errors["title"] = "タイトルは必ず入力して下さい"

    if mode == "create":
        if not line_count:
            errors["line_count"] = "本数を選択してください"
        elif not validate_line_count(line_count):
            line_count = None
            errors["line_count"] = "本数は3～20 の間で選択してください"
        else:
            line_count = int(line_count)
    else:
        line_count = current_line_count if current_line_count else None


    if not amida_items:
        errors["amida_items"] = "アイテムを入力してください"

    if not admin_password:
        if mode == "create":
            errors["admin_password"] = "管理パスワードを入力してください"
    else:
        if not validate_password(admin_password):
            errors["admin_password"] = "管理パスワードは 3 文字以上 20 文字以内で入力してください"
        elif admin_password != admin_password_confirm:
            errors["admin_password"] = "管理パスワードは、上段と下段で同じものを入力して下さい"

    if user_password:
        if not validate_password(user_password):
            errors["user_password"] = "利用パスワードは 3 文字以上 20 文字以内で入力してください"
        elif user_password != user_password_confirm:
            errors["user_password"] = "利用パスワードは、上段と下段で同じものを入力して下さい"

    raw_amida = {
            "title": title,
            "line_count": line_count,
            "option_auto_open": option_auto_open,
            "option_hide_items": option_hide_items,
            "admin_password": admin_password,
            "user_password": user_password
            }

    amida_items = amida_items[:line_count] if line_count else amida_items

    return errors, raw_amida, amida_items
